code/modules/projectiles/gun_system.dm

```diff
@@ -456,24 +458,54 @@
 			action.update_button_icon()
 
 	update_item_state()
-	update_mag_overlay()
 
-///Updates the magazine overlay, this uses the Attachment overlays and is primarily used for extended magazines.
-/obj/item/weapon/gun/proc/update_mag_overlay(mob/user)
+/obj/item/weapon/gun/update_icon_state()
+	. = ..()
+	if(CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_TOGGLES_OPEN) && !CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_CLOSED))
+		icon_state = base_gun_icon + "_o"
+	else if(CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_REQUIRES_UNIQUE_ACTION) && !in_chamber && length(chamber_items))
+		icon_state = base_gun_icon + "_u"
+	else if((!length(chamber_items) && max_chamber_items) || !rounds)
+		icon_state = base_gun_icon + "_e"
+	else if(current_chamber_position <= length(chamber_items) && chamber_items[current_chamber_position] && chamber_items[current_chamber_position].loc != src)
+		icon_state = base_gun_icon + "_l"
+	else
+		icon_state = base_gun_icon
+
+//manages the overlays for the gun - separate from attachment overlays
+/obj/item/weapon/gun/update_overlays()
+	. = ..()
+	//ammo level overlays
+	if(ammo_level_icon && length(chamber_items) && rounds > 0)
+		var/remaining = CEILING((rounds /(max_rounds)) * 100, 25)
+		var/image/ammo_overlay = image(icon, icon_state = "[ammo_level_icon]_[remaining]", pixel_x = icon_overlay_x_offset, pixel_y = icon_overlay_y_offset)
+		. += ammo_overlay
+
+	//magazines overlays
 	var/image/overlay = attachment_overlays[ATTACHMENT_SLOT_MAGAZINE]
-	overlays -= overlay
 	if(!CHECK_BITFIELD(reciever_flags, AMMO_RECIEVER_MAGAZINES) || !length(chamber_items))
 		attachment_overlays[ATTACHMENT_SLOT_MAGAZINE] = null
 		return
 	if(!get_magazine_overlay(chamber_items[current_chamber_position]))
 		return
 	var/obj/item/current_mag = chamber_items[current_chamber_position]
-	overlay = image(current_mag.icon, src, get_magazine_overlay(current_mag))
+	overlay = get_magazine_overlay(current_mag)
 	attachment_overlays[ATTACHMENT_SLOT_MAGAZINE] = overlay
-	overlays += overlay
+	. += overlay
 
 /obj/item/weapon/gun/update_item_state()
-	item_state = "[base_gun_icon][flags_item & WIELDED ? "_w" : ""]"
+	if(ammo_level_item == FALSE)
+		item_state = "[base_gun_icon][flags_item & WIELDED ? "_w" : ""]"
+	else
+		. = item_state
```

Comment:
early return
also on a sidenote why . here?

Reply:
It was like that in the previous proc under eguns so I kept it. I've changed it to it's own var.